File: c++/SplitBinarySearchTree.java
/**
 * Created on 7/24/16.
 * Given a binary search tree and a value,
 * split the binary search tree into two
 * binary search trees whose node values
 * are less/equal and greater than the given
 * value.
 */

public class SplitBinarySearchTree {

    public TreeNode[] splitBST(TreeNode root, int val) {
        if (root == null) return new TreeNode[]{null, null};

        // left and right are the two new
        // tree roots in this call frame
        TreeNode left = null, right = null;
        if (root.val <= val) {
            left = root;
            TreeNode[] nodes = splitBST(root.right, val);
            nodes[0].right = left;
            root.right = nodes[0];
            right = nodes[1];
        } else {
            right = root;
            TreeNode[] nodes = splitBST(root.left, val);
            root.left = nodes[1];
            left = nodes[0];
        }
        return new TreeNode[]{left, right};
    }
}

File: c++/split_binary_search_tree.cpp
/**
* Created on 7/24/16.
* Given a binary search tree and a value,
* split the binary search tree into two
* binary search trees whose node values
* are less/equal and greater than the given
* value. (G)
*/

#include <iostream>

using namespace std;

struct TreeNode {
    TreeNode(int val) {
        this->val = val;
        left = nullptr;
        right = nullptr;
    }
    TreeNode* left;
    TreeNode* right;
    int val;
};

class Solution {
public:
    pair<TreeNode*, TreeNode*> splitBinarySearchTree(TreeNode* root, int val) {
        if (nullptr == root) return {nullptr, nullptr};
        pair<TreeNode*, TreeNode*> roots = {nullptr, nullptr};
        if (root->val <= val) {
            pair<TreeNode*, TreeNode*> p = splitBinarySearchTree(root->right, val);
            roots.first = root;
            root->right = p.first;
            roots.second = p.second;
        } else {
            pair<TreeNode*, TreeNode*> p = splitBinarySearchTree(root->left, val);
            roots.second = root;
            root->left = p.second;
            roots.first = p.first;
        }
        return roots;
    }
};

int main()
{
    TreeNode* root = new TreeNode(3);
    root->left = new TreeNode(2);
    root->left->right = new TreeNode(1);
    root->right = new TreeNode(4);
    root->right->left = new TreeNode(5);
    Solution sol;
    pair<TreeNode*, TreeNode*> roots = sol.splitBinarySearchTree(root, 3);
    cout << roots.first->val << endl;
    cout << roots.second->val << endl;
    return 0;
}
